### DISCUSSION TITLE
Looking for dedicated users willing to test experimental new builds

### DISCUSSION THREAD

**Discussion by mcdope** [mcdope/pam_usb#124] [Show and tell]
Are there any users out there that want to help testing new experimental builds?

I plan to do some huge refactoring for which it would be nice to have people that are familiar with pam_usb. Ideally you would use it in an "non-standard" scenario (like unusual display managers, exotic distributions etc etc) but any testing would be highly appreciated. 

For this I would provide .deb packages, users of other distributions would need to do compile from source from a given branch.

**Comment by mcdope** [comment]


**Reply by mcdope** [reply]
First signed RPM: https://www.dropbox.com/s/nc9kcsxmcm94uof/pam_usb-0.8.2-1signed.x86_64.rpm?dl=0

**Comment by Fuseteam** [comment]
hi hi, i just saw your email! i'm willing to test, once i find me an stick i'll give it a whirl for sure~

**Reply by mcdope** [reply]
Nice, thanks for joining this discussion since I can now just tag when new stuff is there to test. I promise I won't overuse it ;-)

**Reply by Fuseteam** [reply]
haha sure looking forward to it xD

**Comment by mcdope** [comment]
"Beta" (Google Style, not really beta but not proper versioned yet :laughing: ) version of 0.8.3 available at branch `prepare-083`

**Reply by mcdope** [reply]
~~RC1 at https://www.dropbox.com/s/oh5a52v068e8iy9/libpam-usb_0.8.3-RC1_amd64.deb?dl=0~~

@Fuseteam can you give it a go too plz?

**Reply by mcdope** [reply]
~~RC2 (only fixiing an issue with debconf for new install): https://www.dropbox.com/s/vsadwvzg0pj6tv8/libpam-usb_0.8.3-RC2_amd64.deb?dl=0~~

**Reply by mcdope** [reply]
RC3, fixing #175 https://www.dropbox.com/s/war2l6iw6nr01nj/libpam-usb_0.8.3-RC3_amd64.deb?dl=0

**Reply by Fuseteam** [reply]
ooh sure lemme download

**Reply by Fuseteam** [reply]
what should i look out for?

**Reply by mcdope** [reply]
Would be nice if you could test if
a) it works for you on your system (thought thats kinda given xD)
and more importantly, since there where related changes in #175 now, b) try to break it with tmux or whatever comes to your mind

**Reply by Fuseteam** [reply]
alright, imma try it this weekend xd

**Reply by mcdope** [reply]
I assume the silence means either that you had no time, or you found nothing. Anyway - this will be released tomorrow 😆 

Edit: merged & released ^^

**Reply by Fuseteam** [reply]
Yeah sorry been a little too busy, thanks for keeping it up to date! Should i find something when i do have time i'll definitely keep posted

**Reply by mcdope** [reply]
No problem at all, totally understand.

In fact I don't have time myself for this currently xD

**Reply by steelsnake** [reply]
Downloaded and installed - seems good so far on Kubuntu 22.10. Will do some extensive testing and let you know how it goes.

**Reply by mcdope** [reply]
This thread is almost half a year old and that release is published for while now :D

I guess you wanted to reply to this one: https://github.com/mcdope/pam_usb/discussions/124#discussioncomment-5353758?

**Comment by steelsnake** [comment]
I can always test... I have two Kubuntu Jammy installs, one with backports enabled and one without. Plus a few servers that I can just run a KVM on with any version you'd like to test on.

**Reply by mcdope** [reply]
Thanks, you can bet I will come back to that since I hate to put out new releases without user feedback. 

Soon (more or less) there will be test builds for multi-device-per-user-support in example.

**Reply by steelsnake** [reply]
Sure, can absolutely test that. Just let me know what system(s) you want this tested on. Since I'll test on KVM, no danger of destroying anything. :)

**Comment by mcdope** [comment]
New test build! 

This replaces the hardcoded "local" whitelist with a config based one. So that weirdo login managers etc. don't require a patch but can just be added by the user. 

Currently it contains the exact same whitelist as previously. So you should not notice any change at all. 

In the future the default list will be empty and the user needs to enable the entries they need.

.deb, .rpm and .zst at: https://www.dropbox.com/sh/h5qxy3kyr12ymq3/AACKxDpJvIsUdYCXuj0EbYUKa?dl=0

**Reply by mcdope** [reply]
Here I would like to get "exotic tests", like xdm, tmux, lxde etc pp.

**Reply by mcdope** [reply]
@Fuseteam @steelsnake maybe you could test this one for some days? no hurry, but would be really nice.

**Reply by Fuseteam** [reply]
Will try to test the coming weeks :D
(Schedule's a bit flaky still)

**Reply by steelsnake** [reply]
Heh! Oki, replying to the right thingamajiggers now. I hope.

First test results are looking good (on Kubuntu 22.04 LTS and Kubuntu 22.10). I tested remote via SSH:
SDDM works locally, I don't have remote access to it
SSH works, allows local, denies remote
SUDO works, Allows local, denies when using a remote session (i.e. from SSH)
Konsole shell works locally
TMUX works in the two test cases I can think of:
 - Locally started: sudo works from local, denies remote
 - Remotely started: sudo works from local, denies remote

I have not yet experimented with the new session settings, will let you know when I get that done too.

Currently I still have two options from my old config active, I need to see if I can run without them:
```
                <defaults>
                        <option name="unknown_pts_as_local">true</option>
                        <option name="one_time_pad">false</option>
                </defaults>
```
Otherwise the config is stock (aside from the device section for my encrypted USB flashdrive).

**Reply by mcdope** [reply]
Thanks! 

`unknown_pts_as_local` does not exist anymore, safe to remove.

**Reply by mcdope** [reply]
Since steelsnakes extensive test (in fact: he tested more then myself :D kudos lol) didn't found any breakage, and I haven't noticed anything either, I'm already gonna merge this (without putting a new release out yet, more incoming).

**Reply by steelsnake** [reply]
FYI I did test the new deny_remote toggles and they do seem to work with a remote X session. I don't have any system capable of running pure Wayland, can't test that easily. Setting the kde deny_remote to true does make the screensaver not recognize the USB key and also does require entering the password for KDE Discover software installs/updates. I'd consider that normal and expected (and possibly even useful in some scenarios).

On a hunch I also tried VNC (xvnc) and TeamViewer connections to the desktop: These get seen as "local" and allow login with the USB key if it's plugged in. Which makes sense and is probably unavoidable. At least I can't think of a way to detect that.

**Reply by mcdope** [reply]
Nice that you tested that! Always wanted to do it but was too lazy so far tbh.

That KDE behaviour is indeed known and the reason why it's on the whitelist. But I want to improve on that, the goal is to get rid of as much whitelisting as possible.

Could you maybe provide `w` and `pamusb-check --debug` while running via xvnc and teamviewer? Maybe we can actually catch that.

**Reply by steelsnake** [reply]
Sorry for the delay. Kinda as suspected, neither xvnc nor TeamViewer sessions will show any sign of a remote connection. They pretty much just capture the screen and inject keyboard/mouse events, simulating a local user. On the other hand I think it exposed an error in a script somewhere:

```
[src/local.c:344]       Trying to get tty by loginctl
sh: 1: Syntax error: "(" unexpected
```

As far as TeamViewer and VNC go, there doesn't seem to be any way to detect an active incoming TeamViewer or VNC connection short of checking active. `ss -tnp` for example will show a line `users:(("TeamViewer_Desk",pid=406339,fd=14))` when there is an active _incoming_ connection.

Full output from Konsole via an established TeamViewer console (looks the same on xvnc):
```
username@testvm:~$ pamusb-check --debug username
* Authentication request for user "username" (pamusb-check)
[src/local.c:227] Checking whether the caller (pamusb-check) is local or not...
[src/local.c:238]       Checking pid 405205 (pamusb-check)...
[src/local.c:238]       Checking pid 405177 (/bin/bash)...
[src/local.c:238]       Checking pid 405156 (konsole)...
[src/local.c:238]       Checking pid   1555 (/lib/systemd/systemd)...
[src/local.c:238]       Checking pid      1 (/sbin/init)...
[src/local.c:296]       Using DISPLAY :0 for utmp search
[src/local.c:050]       No utmp entry found for tty ":0"
[src/local.c:312]       Trying to get tty from display server
[src/local.c:322]               Failed, no result while trying to get TTY from display server
[src/local.c:327]       Trying to get tty by DISPLAY
[src/local.c:335]               Failed, no result while searching utmp for display :0 owned by user username
[src/local.c:344]       Trying to get tty by loginctl
sh: 1: Syntax error: "(" unexpected
[src/local.c:214]               'loginctl' returned nothing.'
[src/local.c:355]               Failed, no result while searching utmp for tty (null)
[src/local.c:367]       Fallback: Using TTY /dev/pts/1 from ttyname() for search
[src/local.c:053]               utmp entry for tty "pts/1" found
[src/local.c:054]                       utmp->ut_pid: 405177
[src/local.c:055]                       utmp->ut_user: username
[src/local.c:071]               Checking utmp->ut_addr_v6[0]
[src/local.c:071]               Checking utmp->ut_addr_v6[1]
[src/local.c:071]               Checking utmp->ut_addr_v6[2]
[src/local.c:071]               Checking utmp->ut_addr_v6[3]
[src/local.c:075]       utmp check successful, request originates from a local source!
[src/local.c:374] No remote access detected, seems to be local request - allowing.
[src/device.c:038] Searching for "KeychainDrive" in the hardware database...
* Authentication device "KeychainDrive" is connected.
[src/device.c:100] One time pad is disabled, no more verifications to do.
* Access granted.
```

...and output from w (also the same for Teamviewer and VNC):
```
USER     TTY      FROM             LOGIN@   IDLE   JCPU   PCPU WHAT
username tty1     :0               17:54    6:00  16.87s  0.26s /usr/bin/startplasma-x11
username pts/0    :0               17:54    6:00   0.00s  1.67s /usr/bin/kded5
username pts/1    :0               17:55    0.00s  0.02s  0.01s w
username tty2     -                17:54    6:00   0.05s  0.03s -bash
```

**Reply by mcdope** [reply]
Hmm that sucks. For TV I almost expected it but thought xvnc would maybe at least show up in the process chain. 

Guess that really needs actually checking if these processes have active incoming connections.

**Reply by steelsnake** [reply]
xvnc might be due to the way I've invoked it:
```
nohup /usr/bin/x11vnc -auth guess -forever -display :0 -loop -noxdamage -repeat -rfbauth /etc/vnc/passwd -rfbport 5900 -shared 1>/dev/null 2>&1 &
```
There also is a way to actually run X11 as a headless VNC server, but I haven't done that in a very long time. But in that case all connections would be remote, making USB authentication pretty moot I think. Not to mention that this kind of configuration ends up being multi-seat, akin to a terminal server. I'd imagine that would pose other challenges too.

**Reply by steelsnake** [reply]
Forgot: With TeamViewer you could just check whether `ss -ntp` returns a line containing `TeamViewer_Desk` as mentioned above. If you like I can see if there's other ways to do this. Netstat should also work, not sure if `ss` or `netstat` is more commonly installed these days.

**Reply by mcdope** [reply]
I would appreciate any testing and information on improving pam_usb, esp. if it's about security. And this very much is. 

Ofc the best would be if we can detect if the current request is local, but knowing there is SOME remote connection to the session like with vnc/teamviewer is also already an improvement. I want to prevent any miss-authentication when deny_remote isn't disabled, even if it maybe means denying really local requests just cuz there is some VNC/TV conn. (And yes, maybe this is a bit too much love for detail - but hell, this is a PAM module after all)

But I don't want to ask too much for your time. If you're bored and have the time for it, sure - test some more and if you find something it will maybe be useful. `ss -ntp` i.e for sure will helpful and i'm almost certain that will end up in the code. But really: only if you can spare the time. My comfort / lazyness shouldn't be your reason ^^

Edit: what I will for sure test very soon: running via RDP. I have a local proxmox server which hosts `cura` in a debian xfce LXC via RDP,. Gonna re-use that to test "rdp-to-login".

Edit 2: Of course any help on improving pam_usb will be documented in the changelog - whatever that is worth

**Reply by steelsnake** [reply]
I'll see if I can figure out a reliable way to detect VNC short of just checking active connections. The other big question would be Wayland, I think.

No need to give any credit anywhere - I generally much prefer to just help because helping is the right thing to do :)

**Reply by mcdope** [reply]
> xvnc might be due to the way I've invoked it:
> 
> ```
> nohup /usr/bin/x11vnc -auth guess -forever -display :0 -loop -noxdamage -repeat -rfbauth /etc/vnc/passwd -rfbport 5900 -shared 1>/dev/null 2>&1 &
> ```
> 

How exactly do you run this, manually, via some service file, something else? Wanna replicate the setup for testing.

**Reply by steelsnake** [reply]
Manually, I have a `vnc.sh` bash script that I invoke whenever I need VNC service (with a handy "STOP" parameter to stop it later):
```
#!/bin/bash

if [[ $1 == STOP ]]; then
        killall x11vnc
else
        nohup /usr/bin/x11vnc -auth guess -forever -display :0 -loop -noxdamage -repeat -rfbauth /etc/vnc/passwd -rfbport 5900 -shared 1>/dev/null 2>&1 &
fi
```
You'd need to generate/store the password with `x11vnc -storepasswd MyPasswd /etc/vnc/passwd`

**Reply by mcdope** [reply]
> Forgot: With TeamViewer you could just check whether `ss -ntp` returns a line containing `TeamViewer_Desk` as mentioned above. If you like I can see if there's other ways to do this. Netstat should also work, not sure if `ss` or `netstat` is more commonly installed these days.

Does this also work for x11vnc? Could you try? I currently cant setup VMs - out of storage :D

**Comment by mcdope** [comment]
There is a new release upcoming (finally). Even if it's just a minor version it's a kinda important update since it includes fixes for stuff with crash (like in segfault :D) or at least "wont work" potential.

If you want a preview: build from this branch / PR -> https://github.com/mcdope/pam_usb/pull/218

Prebuilt packages available at -> https://www.dropbox.com/scl/fo/kmll4q7tz3z0agbk4q30x/h?rlkey=da59dbbovgv3ip1qch3hy6lz4&dl=0 (still uploading right now)

**Reply by mcdope** [reply]
@Fuseteam you are the one with tmux if I remember correctly? Your feedback would be esp. nice since the memory fixes affect tmux handling. For me it still works like expected, but you know tmux better than me.

**Reply by steelsnake** [reply]
Hey there! Sorry for the late reply, I was a bit busy. I just installed the package on both of my Ubuntu boxes (one on 23.04 the other on 22.04 LTS). So far, looks good. I'll report back Sat or Sun to let you know if anything blows up.

I do use tmux all the time btw, started with a simple systemd service (attached, added .txt so git doesn't freak out) 
[tmux.service.txt](https://github.com/mcdope/pam_usb/files/13834156/tmux.service.txt)
...and then attaching when needed via `tmux attach -t default`

Thanks for the update.

**Reply by steelsnake** [reply]
Alright, a bit later than expected thanks to pesky RL and my flash drive died. LOL

I didn't notice anything out of the ordinary, seems to work as intended. No crashes, nothing.

**Reply by mcdope** [reply]
Dont worry about however long it takes, I'm very thankful for any feedback :D 

History has proven that "works on my machine" isnt the best indicator :D So: thanks for testings! I also haven't noticed any issues. Even tried sudo bash in tmux with tmux on top with sudo.... etc pp ^^

Edit: these test builds are btw now published as 0.8.4

**Reply by Fuseteam** [reply]
hey i now see this, i'll test when i have time soon™ too

**Comment by mcdope** [comment]
@Fuseteam @steelsnake and everyone interested: experimental new test builds at https://www.dropbox.com/scl/fo/y5tha7yhx18uo35gal1rv/AFIjBlTnmd4RfkrjHCwv6rA?rlkey=boqwcnbrgfjk34dbtlrpcafk9&dl=0, source is in branch `fix-some-mem-stuff` / #229 .

These are `master` plus changes proposed by a new coding AI, mainly related to memory and string handling. I expect this to crush some exotic bugs. 

What I need you to test is just "did it break anything?" :grin:  I couldn't notice any issues but I want to be sure before merging.

**Reply by steelsnake** [reply]
Works fine on my home system. I can test it on my "work" PC on Monday.

Thanks for all your work, much appreciated!

**Reply by mcdope** [reply]
I also couldn't find any issues so far, tried on desktop and laptop. 

Will continue to use it for some days before merging though.

**Reply by steelsnake** [reply]
Okay, seems to work on my other PC as well. No issues so far, but only tested for a little bit.

- Local: Login works. sudo works
- Remote:
  * SSH properly denies
  * sudo properly denies
  * VNC+TeamViewer allows (makes sense as it's not recognized as remote)
  
Both VNC and TeamViewer I think are simply beyond the scope of pam_usb, so not really worried about that. There's just too many different remote access solutions out there.

**Reply by mcdope** [reply]
For TV there is actually a way, but I first need to find a way to test for VNC too. Esp. since I guess VNC is more common on *nix then TV.

**Comment by mcdope** [comment]
For the brave ones: there is now a first preview build offering support for multiple devices per user (#31). You can find the code in PR #177. 

However, well possible there are still some bugs left. It's only VM-tested for short time now. Also it's still full of debugging output that will vanish before it will be released.

You can find signed .deb and .rpm packages in this Dropbox folder: https://www.dropbox.com/scl/fo/piormhnmfnqquubspwu4o/AM_yFraRt9H6qUt3umqDY8E?rlkey=o8idyui7cf1klmteqh5m35ov2&dl=0

**Be aware: If you install this build AND configure multiple devices, your config wont be backwards compatible anymore afterwards. To downgrade to another version you will have to manually remove all but one device from your user section in the config.**

**Reply by steelsnake** [reply]
Seems to work fine I just installed it on a second PC. I did however NOT use a multiple device setup. I'll try that sometime this weekend.

**Reply by mcdope** [reply]
There is now a new build in that dropbox folder, having improved agent support for multiple devices. This is now pretty much complete, except the still left debug outputs.

I couldn't find any issues so far in either the "normal" stuff or the new "multidevice" stuff.

**Reply by mcdope** [reply]
And another new build in the folder. If nobody notices/reports any bugs with it this is the final version of that feature.

To me it seems fine, tested it on VM, Desktop and Laptop and couldn't notice any issues.

**Reply by mcdope** [reply]
This feature is now merged and available in `master`, next nightly build will contain it too.

Release will be as 0.8.5 some when in the future.

**Reply by steelsnake** [reply]
As near as I can tell the latest version does seem to work fine. I've installed it 2 days ago and don't see any issues so far. If I can test anything specific, do let me know please.

**Comment by mcdope** [comment]
You could try adding a second device for your user.

But I don't expect any issues (as long as you don't try adding more then
ten devices per user :D)


Am Sa., 20. Juli 2024 um 17:57 Uhr schrieb steelsnake <
***@***.***>:

**Reply by steelsnake** [reply]
Just tried that, does seem to work fine on both of my test machines.

Unrelated, but I found a KDE bug while testing: If you lock the screen while the USB key is connected, then unplug it while the screen is locked, you can still unlock the screen without password. KDE seems to simply check the credentials on lock rather than unlock. Just an FYI in case anyone runs across that.

**Reply by mcdope** [reply]
Are you sure you didn't had a second device connected at that time? 

If so: on what distribution and with which KDE version? Would like to replicate for checking.

**Reply by steelsnake** [reply]
Absolutely sure :)

Kubuntu 24.04

System Info:
Operating System: Kubuntu 24.04
KDE Plasma Version: 5.27.11
KDE Frameworks Version: 5.115.0
Qt Version: 5.15.13
Kernel Version: 6.8.0-38-generic (64-bit)
Graphics Platform: Wayland

Not tested on Qt 6

Test:
- Connected device: Bus 001 Device 008: ID 154b:00f3 PNY USB 3.0 FD
- Lock screen (don't log off, just lock), wait until "lock" screen with "unlock" button appears
- Unplug device, wait a few secs just in case
- Can unlock screen just by pressing the "Unlock" button
![image](https://github.com/user-attachments/assets/885f966c-32b7-413b-8190-b5b51e85d235)

Comparison, locking without USB device:
- Unplug device before locking
- Lock screen
- As expected, asks for password
![image](https://github.com/user-attachments/assets/7a14c242-699a-48d0-a2a2-a80337541098)

**Reply by Fuseteam** [reply]
I just realized we could approach this differently; unplugging the usb key does not invalidate the authentication.

in other words locking the screen while the key is plugged in, allows you to unlock without password. Unplugging the key doesn't do anything

**Reply by mcdope** [reply]
Unplugging the key would work, if the agent is configured to lock the screen. Since then the device is already disconnected this behavior shouldn't occur. Haven't tested it yet. 

Will be on summer holiday for the next three weeks btw

**Reply by Fuseteam** [reply]
enjoy your vacation~

**Reply by Fuseteam** [reply]
@steelsnake is the agent configured to lock the screen for you?

**Reply by steelsnake** [reply]
How would i go about doing that? I only configured things according to the pam_usb instructions, nothing special.

**Reply by Fuseteam** [reply]
not sure, but i think it might be configured to do that, which why unplugging doesn't lock it.
So i guess it's less about unplugging not locking it, but locking it while the key being connected, not locking it :thinking:

**Reply by Fuseteam** [reply]
this might help: https://github.com/mcdope/pam_usb/wiki/Getting-Started#agent

**Reply by mcdope** [reply]
> How would i go about doing that? I only configured things according to the pam_usb instructions, nothing special.

@Fuseteam is right on that one, see that Wiki page and esp. the linked one (https://github.com/mcdope/pam_usb/wiki/Configuration#agent). You basically need to a) install the service (only if you haven't installed from package) b) configure the agent for your user c) enable the agent.

I can't tell you the correct way to set it up for KDE though, last time I've really used KDE was on SuSE 6.2 :laughing:

**Reply by Fuseteam** [reply]
thought you were on vacation :joy:

**Reply by mcdope** [reply]
That means to me: ignoring work and taking time for private stuff :D Yesterday that meant swimming with my daughter, today this means private projects haha :D 

Will do only day-trips this year for $wierdLifeReasons ^^

**Reply by Fuseteam** [reply]
haha makes sense

**Reply by steelsnake** [reply]
> not sure, but i think it might be configured to do that, which why unplugging doesn't lock it. So i guess it's less about unplugging not locking it, but locking it while the key being connected, not locking it 🤔

So, after doing some research and adjusting the config accordingly, locking and unlocking by unplugging/plugging the USB stick works. It does not change the behavior of the KDE lock screen though. If you lock the screen, unplug the USB stick, the "lock" command does trigger but the lock screen still keeps the "Unlock" button, no password nor USB drive required.

For reference, the "proper" command to lock is either `qdbus org.freedesktop.ScreenSaver /ScreenSaver Lock` or `loginctl lock-session` (the latter is, AFAIK, systemd specific).

To unlock, qdbus seemingly has no way to do this, I don't know why. Only `loginctl unlock-session` works.

**Reply by mcdope** [reply]
> > not sure, but i think it might be configured to do that, which why unplugging doesn't lock it. So i guess it's less about unplugging not locking it, but locking it while the key being connected, not locking it 🤔
> 
> So, after doing some research and adjusting the config accordingly, locking and unlocking by unplugging/plugging the USB stick works. It does not change the behavior of the KDE lock screen though. If you lock the screen, unplug the USB stick, the "lock" command does trigger but the lock screen still keeps the "Unlock" button, no password nor USB drive required.
> 
> For reference, the "proper" command to lock is either `qdbus org.freedesktop.ScreenSaver /ScreenSaver Lock` or `loginctl lock-session` (the latter is, AFAIK, systemd specific).
> 
> To unlock, qdbus seemingly has no way to do this, I don't know why. Only `loginctl unlock-session` works.

Yeah, if you lock the screen manually that's "expected behaviour" kinda. The interesting bit would be, does the same happen with the auto-lock by the agent? I would almost bet it doesn't.

Thanks for the commands, gonna update the wiki soon ™️

**Reply by steelsnake** [reply]
Just tried it, I kinda didn't think about trying THAT. LOL... braindead. Anyway, as you said, locking by unplugging does properly lock the system, with password prompt. As expected, it also unlocks properly when plugging the USB stick back in.

**Reply by mcdope** [reply]
Thanks for trying. Than it is for sure an "inverted variant" of the lightdm issue.

**Reply by Fuseteam** [reply]
lovely, fun enough gnome doesn't appear to have this; it even seems to briefly pop up the password prompt while the key is plugged in

**Reply by mcdope** [reply]
Just took a look at the src of pam_poldi for use with GPG smartcards. They just block the chain until $event (Card inserted, PIN entered) happens or a timeout triggers. So basically the same like my suggested lightdm workaround (#221).

I never used SmartCards on a PC, but I assume you have to enter the PIN on each usage - so this wouldn't affect the KDE-Lockscreen in the same way like pam_usb.

This research result is also an argument to stop doing workarounds kinda imho.

**Reply by Fuseteam** [reply]
> This research result is also an argument to stop doing workarounds kinda imho.

yeah especially if there is a general way

**Comment by mcdope** [comment]
Lovely, thanks for the detailed instructions - much appreciated

Am Mo., 22. Juli 2024 um 20:57 Uhr schrieb steelsnake <
***@***.***>:

**Comment by mcdope** [comment]
Hmmm when I see those screens, and re-read
https://github.com/mcdope/pam_usb/issues/221, I smell the same behaviour.
Or in other words: your diagnose of " KDE seems to simply check the
credentials on lock rather than unlock" is most likely absolutely on point.

I bet the lockscreen starts the PAM chain directly, at which point pam_usb
will (still) see the device and allow the PAM request.

But unlike for `lightdm`, #221 wouldn't fix your issue. For that we need
something like an "ignore the first request" option (which obviously should
be set on `<service>` level instead of globally. Which would also kinda
solve #221, even if a bit more annoying for the user...

I guess if you wouldn't lock manually but let the agent do it when the
device is unplugged this wouldn't occur.

Am Mo., 22. Juli 2024 um 21:01 Uhr schrieb Tobias Bäumer <
***@***.***>:

**Reply by steelsnake** [reply]
My view on this is that pam_usb shouldn't try to fix a KDE issue. It's good to be aware of it, but in the end it's a broken KDE authentication flow. It should only ever authenticate on unlock, not on lock. I _think_ I know why they do it the way they do, but it's still broken: As the lock screen can be themed, they likely provide the theme clue on initial display, i.e. on lock.

**Reply by mcdope** [reply]
Hmmm, also a very valid point which I kinda share. 

But then again, I guess most users would count this as bug instead of "working like intended". 

But then again, the workaround suggested by me (ignoring first request) can also be easily counted as "bug" instead of "feature" 😆.

Let's make a a poll out of this: react with:
- 👎 for NOT workarounding issues like this (would also imply closing #221)
- 👍 for ADDING WORKAROUNDS for such behaviour.

I agree with steelsnake, but don't wanna be a dictator for the project so I will also not vote ^^. I will check/count the results on 4th august at some time.

**Reply by mcdope** [reply]
Considering the participation it seems my attempt at not being a dictator resulted in you being one haha. 

Funny, but kinda expected. Was worth the try still, vote will continue to run till the 4th ^^

**Reply by steelsnake** [reply]
Haha. Sorry!

**Reply by Fuseteam** [reply]
dang i need to be more active on github lolrip

**Reply by Fuseteam** [reply]
i think for the kde one, we should get touch with the kde maintainer to see if we can solve it upstream first, for lightdm i'm not sure who the current maintainer is tho :thinking:

**Reply by mcdope** [reply]
I doubt that KDE or lightdm would adjust something for us. Also I wonder how both react to a) fingerprint readers b) to smartcards, or better said: how do these modules handle it.

But I don't have hardware to test either.

Edit: my guess would be that stuff using biometry or physical stuff like smartcards usually just blocks the chain till they have a result. But for one that would break password-login in case of failure but also: this is just not how PAM Is supposed to work except I understood it incorrectly (which is totally an option :D)

Edit 2: Biometry stuff only applies to login, smartcard also to lock/unlock I guess.

**Reply by Fuseteam** [reply]
i'd say 'asking is free' worse case we learn more about how their inner working work :D

**Reply by mcdope** [reply]
Considering the participation so far it's kinda unlikely we will get any more votes on this topic still today... So I call it a day, the vote is over.

There will be NO (further) workarounds for "broken" / "wierd" PAM clients like `lightdm` or `kscreenlocker` added.

**Reply by Fuseteam** [reply]
sounds good

**Comment by mcdope** [comment]
It's "that time of the year again" - I need dedicated tested for a new build. Please try as much as you can and just see if everything still works for you. 

This is a security related update, so I won't give details about changes till the final update is released. Same applies for sourcecode of this update, will be released after/with the update for security purposes.

You can find the testing builds at https://www.dropbox.com/scl/fo/dc8blo5sv3eetneagi6bs/AL_SwIdDBd4eg8vZHST2MVU?rlkey=h9cbgh5a1gelxqk6erk1qobk9&st=hlzet1ms&dl=0

**Reply by mcdope** [reply]
@Fuseteam @steelsnake

**Reply by Fuseteam** [reply]
will take a look when i have a change :saluting_face:

**Reply by mcdope** [reply]
You were the one with tmux, right? That would be esp. interesting.

**Reply by Fuseteam** [reply]
oh yeah i remember that :grin:

**Reply by steelsnake** [reply]
Downloaded the Ubuntu .deb, installed it, going to test :)

**Reply by steelsnake** [reply]
Tested and working so far:
* Local sudo works
* SDDM login works
* TMUX locally: sudo bash tosses this error: `Could not get properties: Unknown object '/org/freedesktop/login1/user/self`.`
* TMUX remotely correctly identifies the session as remote and denies the request
* SCREEN locally: Access denied (no further error presented, just access denied)
* SCREEN remotely: Correctly identifies session as remote and denies the request

This is on Kubuntu 25.10, Wayland

**Reply by mcdope** [reply]
> Tested and working so far:
> 
>     * Local sudo works
> 
>     * SDDM login works
> 
>     * TMUX locally: sudo bash tosses this error: `Could not get properties: Unknown object '/org/freedesktop/login1/user/self`.`
> 
>     * TMUX remotely correctly identifies the session as remote and denies the request
> 
>     * SCREEN locally: Access denied (no further error presented, just access denied)
> 
>     * SCREEN remotely: Correctly identifies session as remote and denies the request
> 
> 
> This is on Kubuntu 25.10, Wayland

Very interesting, usually the bugs are in deny_remote and not in the local part :D Will debug that tomorrow.

**Reply by steelsnake** [reply]
Holler if I can provide more info.

**Reply by mcdope** [reply]
tmux locally works fine for me with that build. Can you give more details on how exactly you tested it? I tried by simply running tmux and then invoking pamusb-check --debug `whoami`.
<img width="2560" height="1408" alt="image" src="https://github.com/user-attachments/assets/4ee03e27-4eb9-4d10-9187-2ad3e0d1b35a" />

**Reply by mcdope** [reply]
Screen works as well for me.
<img width="1018" height="813" alt="image" src="https://github.com/user-attachments/assets/333452c9-ca1e-451b-8df5-7ddf8e3719bb" />

**Reply by KeyofBlueS** [reply]
Really sorry for the delay.

My main use of pamusb is for sudo and pkexec authentication, and so far I haven't encountered any issues with the testing builds.

If there are any specific tests you'd like me to run, just let me know, I'm happy to help.

**Reply by mcdope** [reply]
Thanks! Well, if it works for you that's already enough testing. Just needed to be sure it doesn't break stuff. 

Then I only need to wait for Github to assign the CVEs before merging/announcing/releasing.

**Reply by KeyofBlueS** [reply]
Sorry, I forgot to mention an important detail earlier.

While everything is working on my side, I had to modify `/etc/pam.d/common-auth` to get pkexec authentication working properly. Specifically, I placed:

`auth	sufficient	pam_usb.so`

at the beginning. Without this change, a password prompt dialog appears instead.

sudo, on the other hand, works fine without any additional changes.

**Reply by mcdope** [reply]
That's actually intended, it doesn't include the recent fix to get polkit working again but is based on plain 0.8.6 instead.

**Reply by mcdope** [reply]
This update is released as 0.8.7. Please update every machine you use pam_usb on immediately!

**Reply by Fuseteam** [reply]
sorry for being slow this time around :joy:

**Reply by steelsnake** [reply]
> > Tested and working so far:
> > ```
> > * Local sudo works
> > 
> > * SDDM login works
> > 
> > * TMUX locally: sudo bash tosses this error: `Could not get properties: Unknown object '/org/freedesktop/login1/user/self`.`
> > 
> > * TMUX remotely correctly identifies the session as remote and denies the request
> > 
> > * SCREEN locally: Access denied (no further error presented, just access denied)
> > 
> > * SCREEN remotely: Correctly identifies session as remote and denies the request
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is on Kubuntu 25.10, Wayland
> 
> Very interesting, usually the bugs are in deny_remote and not in the local part :D Will debug that tomorrow.

Sorry for the late reply, I got distracted by work. tmux is just called as user for me, and then just `sudo bash`:
See store for details.

$ sudo bash
* Authentication request for user "jenni" (sudo)
Could not get properties: Unknown object '/org/freedesktop/login1/user/self'.
Could not get properties: Unknown object '/org/freedesktop/login1/user/self'.
* Access denied.
[sudo: authenticate] Password:

It's not a big deal either, I rarely use sudo within tmux :) I haven't tried it with 0.8.7 yet, will try tomorrow, also will try on Kubuntu 24.04 - I don't have 26.04 yet for testing.

Thanks for all the work and the continued development. MUCH appreciated.

**Reply by mcdope** [reply]
@steelsnake Seems to be a problem of your installation. Just installed a fresh Kubuntu 25.10 install, updated it, installed pam_usb and tested both. Both are working for me. See https://github.com/mcdope/pam_usb/issues/293#issuecomment-4390843770

Would be nice if you could add `<option name="debug">true</option>` to your <defaults> config and try again, that would maybe point into the right direction. But let's continue the discussion about that in the linked ticket.

**Comment by mcdope** [comment]
New release, 0.9.2, arrived. Again many security fixes.

Please try a) if it broke something and b) if you can break it :D

**Reply by steelsnake** [reply]
Will do when I can. I'm still in the hospital, but hopefully out on Tuesday or Wednesday.

Thanks for all your work, sorry I was out of commission for a while.

**Reply by mcdope** [reply]
Get well soon :)
